### ISSUE TITLE
[css-borders-4] add "radius/corner-smoothing" option to `border-radius` for squircles/superellipse

### ISSUE THREAD

**Issue by jsnkuhn** [w3c/csswg-drafts#10653]
https://drafts.csswg.org/css-borders-4/

Idea here is to add sqircles/superellipse as a option for `border-radius` or a separate property maybe "radius-smoothing" instead of including `squircle` as a shape option for "corner-shape".

Gives the benefit of being able to be used with `corner-shape` later to round the corner joints. 

`round` would no longer need to be an option for `corner-shape`.

This would more closely line up with how vector software and iOS already approach squircles/superellipses

----------------------------------

Figma's radius-smoothing https://help.figma.com/hc/en-us/articles/360050986854-Adjust-corner-radius-and-smoothing
https://www.youtube.com/watch?v=52pPP4HjieM

Has fine grained control over the smoothing. Also points out that 60 is the corner smoothing value used by iOS.

Swift UI uses "continuous" as a keyword for this: https://developer.apple.com/documentation/swiftui/roundedcornerstyle/continuous

Looks like this is an all or nothing thing in Swift UI. So either all corners are squircle or none are, no ability to mix. (Honestly not sure that I've ever seen an example of say a top right squircle and a bottom left straight round corner though so probably fine?)

Also looks like "continuous" is a binary keyword with no abiity to control the amount of smoothing. Not sure if that would be desirable? 

If the main point of squircles in CSS is for things to better line up with SwiftUI's iOS styling than maybe this isn't an issues. Although maybe they change this in the future??

Saw a comment in a YouTube video that said "continuous" might now be the default styling for cornerRadius? Would need to be confirmed.

------------------------------------------------------------------------------

Worth looking into what other vector drawing software might call this. 2 article deepdives:

On a plug-in for adobe illustrator:
https://bootcamp.uxdesign.cc/smooth-corner-rounding-in-adobe-illustrator-94003145a7bf

On implementing smoothing in figma:
https://www.figma.com/blog/desperately-seeking-squircles/

**Comment by smfr** [comment]
Dup of https://github.com/w3c/csswg-drafts/issues/6296 ?

**Comment by jsnkuhn** [comment]
This is an alternate proposal to add the smoothing behavior of squircles as an option to the `border-radius` property. And therefore not having to add `squircle` as a keyword to `corner-shape` which is the purpose of #6296

**Comment by didier** [comment]
This would make the web feel so much nicer! Consider this my vote.

**Comment by mohsen1** [comment]
Somehow related:
I tried implementing this with `clip-path` or `image-mask` + container queries. Given that I can get width and height of the element using container queries (`100cqw` and `100cqh`) I could potentially compute the `path()` function using a bunch of CSS `calc()` calculations using width, height and border-radius value. However:

* `path` does not support CSS variables in its value
* `url` (for image-mask`) also does not support CSS variables in the value.

I understand that the value inside of the `url()` is not computed (for security reasons perhaps) but would it be possible to let users use CSS variables in the `path()` function? This will unlock almost any shape (dynamically computed) rather than this very specific use-case.

**Comment by smfr** [comment]
You're asking for the [`shape()` function](https://drafts.csswg.org/css-shapes-2/#shape-function).

**Comment by noamr** [comment]
I think this can be closed now?

**Comment by jsnkuhn** [comment]
https://github.com/w3c/csswg-drafts/issues/10993#issuecomment-2622578937

So, If I'm reading the resolution here correctly. What has been resolved is to do exactly the opposite of what I am proposing here.

The resolution is to make a "superellipse" keyword for corner-shape instead of adding "corner-smoothing" to `border-radius` therefor closing the door to later potentially using border-radius to round the joints of the other "corner-shape" options with smoothed circles?

**Comment by tabatkins** [comment]
> The resolution is to make a "superellipse" keyword for corner-shape

Ah, no, it was to add superellipse() *function*, with the parameter giving the ellipse exponent (or a scaled version of it).  The nice thing about the superellipse shape is that all the major suggested shapes are exactly represented by certain values of the ellipse exponent.

**Comment by jsnkuhn** [comment]
OK, let me try this: my understanding is the superellipse() function is what the browser vendors will be using behind the scenes to create the different shapes. At the moment, at least, devs would have `round`, `scoop`, `bevel`, `notch`, `straight`, `squircle` options.

I get that this is a working draft and that's probably what's tripping me up a bit here. So I'm guessing that ultimately all the superellipse() function references will only be in the "Interpolating corner shapes" section?

------------------------------------------------------------

_This_ issue piggy backs on a comment put forward by @LeaVerou here 
https://github.com/w3c/csswg-drafts/issues/6980#issuecomment-1221512890

>Idea: There are far more use cases for irregular polygons with some rounding on their corners, than for scoop and notch all together. So I wonder if perhaps we need a separate, orthogonal property for the bevel aspect of these corners, with border-radius continuing to apply for _their_ corners.

So there would be a something like this but with notch instead of bevel:

```css
corner-shape:  square square notch square;
corner-size: 0 0 40px 0;
/* A `corner-size` property would be needed to have `border-radius` free to round the joints of each corner */

border-radius:  20px;
corner-smoothing:  ???;
/* The `corner-smoothing` would turn the border-radius rounding into squircle rounding */
```

![Image](https://github.com/user-attachments/assets/884050dc-c8d8-4d89-8196-cea397b596ac)

**Comment by noamr** [comment]
> OK, let me try this: my understanding is the superellipse() function is what the browser vendors will be using behind the scenes to create the different shapes. At the moment, at least, devs would have `round`, `scoop`, `bevel`, `notch`, `straight`, `squircle` options.
> 
> I get that this is a working draft and that's probably what's tripping me up a bit here. So I'm guessing that ultimately all the superellipse() function references will only be in the "Interpolating corner shapes" section?
> 
> _This_ issue piggy backs on a comment put forward by [@LeaVerou](https://github.com/LeaVerou) here [#6980 (comment)](https://github.com/w3c/csswg-drafts/issues/6980#issuecomment-1221512890)
> 
> > Idea: There are far more use cases for irregular polygons with some rounding on their corners, than for scoop and notch all together. So I wonder if perhaps we need a separate, orthogonal property for the bevel aspect of these corners, with border-radius continuing to apply for _their_ corners.
> 
> So there would be a something like this but with notch instead of bevel:
> 
> corner-shape:  square square notch square;
> corner-size: 0 0 40px 0;
> /* A `corner-size` property would be needed to have `border-radius` free to round the joints of each corner */
> 
> border-radius:  20px;
> corner-smoothing:  ???;
> /* The `corner-smoothing` would turn the border-radius rounding into squircle rounding */
> ![Image](https://github.com/user-attachments/assets/884050dc-c8d8-4d89-8196-cea397b596ac)

I wouldn't implement this shape as a single corner, but rather as a `polygon()` with rounded corners or as a `shape()`

**Comment by tabatkins** [comment]
Right, there will be a lot of overlap (and people coming up with incredible hacks that we didn't consider when designing the feature is always great, so long as they depend on specified behavior and not browser bugs). But that doesn't mean we need both of them to be able to do everything the other can.

border-radius shapes the four corners of the box. Giving it the ability to *create* new corners *and then shape those too* seems like too far, given the design of the property. Cases like your example are better handled by `polygon()` (in `clip-path`, or perhaps `shape-inside/outside`), which *is* designed to be able to create an arbitrary number of corners and round them (see <https://drafts.csswg.org/css-shapes/#funcdef-basic-shape-polygon>)

**Comment by jsnkuhn** [comment]
I'm not suggesting there be complete parity between `corner-shape` and css-shape functions. I can however see a potential frustration from devs with `corner-shape` coming SO close to what they are looking for but not quite getting there ("if only i could round all those those corners").

I'd also like to push back at the "If `corner-shape` doesn't quite make it they'll just use css-shape functions etc.". I believed this for a long time because this is how it's always worked for _me_ but since starting answering CSS question in Kevin Powell's discord a few years ago I've run into so many cases where instead of using the slightly more complicated CSS property that is fit for the job (polygon() fits this bill) the devs respond with: "Oh, that looks too complicated. I'm just going to change the design."

**Comment by noamr** [comment]
I created an explainer for `corner-shape`, and addressed this issue: https://github.com/noamr/explainers/blob/main/corner-shape-explainer.md

**Comment by shannonmoeller** [comment]
Saw [an article about `corner-shape`](https://frontendmasters.com/blog/drawing-css-shapes-using-corner-shape/) and immediately wanted to know how to make rounded shapes given `border-radius` was being overloaded to now mean `corner-size`.  A giant +1 to everything @jsnkuhn has said above. The text-flow concern is noted, but we already have to [work around that issue with border-radius](https://codepen.io/shannonmoeller/pen/XJbyQre).

**Comment by zpaolo11x** [comment]
There is a fundamental difference in how corner-shape works with respect to the "radius smoothing" proposal: if you set a border-radius of 50px and a squircle corner-shape, the squircle curve will occupy a 50px x 50px area, thus producing a perceived radius that is much smaller than 50px. 

This causes two issues: falling back to a normal "circular" radius (for browsers that don't support corner-shape) will require some trickery to scale the border-radius value, and in general terms it won't fit the way "smooth corners" are used in vector design software where you set the radius and the smoothing without affecting the general look of the radius. 

Oh and it will prevent creating pill shaped rectangles with "smooth" corners, which is often used in controls and buttons.
